dvc/repo/artifacts.py

```diff
@@ -50,18 +59,31 @@ def read(self) -> Dict[str, Dict[str, Artifact]]:
             dvcfile,
             dvcfile_artifacts,
         ) in self.repo.index._artifacts.items():  # pylint: disable=protected-access
-            # read the artifacts.yaml file if needed
-            if isinstance(dvcfile_artifacts, str):
-                dvcfile_artifacts = ArtifactsFile(
-                    self.repo,
-                    os.path.join(os.path.dirname(dvcfile), dvcfile_artifacts),
-                    verify=False,
-                ).load()
-            if not dvcfile_artifacts:
-                continue
             dvcyaml = relpath(dvcfile, self.repo.root_dir)
             artifacts[dvcyaml] = {}
             for name, value in dvcfile_artifacts.items():
                 check_name_format(name)
                 artifacts[dvcyaml][name] = Artifact(**value)
         return artifacts
+
+    def add(self, name: str, artifact: Artifact, dvcfile: Optional[str] = None):
+        # this doesn't update it "in place": self.read() won't return the updated value
+        if not dvcfile or dvcfile[-len("dvc.yaml") :] != "dvc.yaml":
+            dvcfile = os.path.join(dvcfile or "", "dvc.yaml")
+        dvcfile_abspath = os.path.join(self.repo.root_dir, dvcfile)
+        Path(os.path.dirname(dvcfile_abspath)).mkdir(exist_ok=True, parents=True)
```

Comment:
I would move this into `_update_project_file`

Reply:
We should not create directories, it's caller's responsibility.

Reply:
@skshetry  - can you explain please? It's not clear to me what you're referring to . 
is "caller" being whoever calls/uses `Repo`? Or `Repo` expected to do this but not `Artifact` object (and friendts)? Is there some agreement that specific levels of the code (or API) have / don't have FS side effects?
Probably because I'm unfamiliar but the comment didn't help me understand better 😄

Reply:
If you pass an invalid path, it should fail, not create it. The API should do the least that's expected from it to work to have less surprises. 

Also, the API is not designed to create artifacts in random dvc.yaml file. Most of the time, it's expected to be in the root of the dvc repo or in your current working directory. 

I guess it's following POSIX-y design rule: do one thing and do it well, in the same way `cp`/`mv` don't create directories.

Reply:
oh, you're not talking about "code" callers in DVC, but DVC api as a whole, ok

Reply:
I got some questions about high level behavior that may relate to those type of choices (adding a model to a non dvc initialized repo?) but it's out of scope for this comment thread I believe